### PR TITLE
Adjust DateHistogram's bucket accounting to be iteratively

### PR DESCRIPTION
**Pull request by martijnvg**
Adjust DateHistogram's consumeBucketsAndMaybeBreak to be iteratively during reduce instead accounting all buckets at the end of the reduce.

In case of many non-empty buckets accounting the number of buckets at the end of the reduce may be too late. Elasticsearch may already have failed with an OOME. This change changes the accounting to happen iteratively during the reduce for non-empty bucket.

Note that for empty buckets accounting of the number of buckets already happens iteratively.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalDateHistogram.java [hunk 1]
```diff
@@ -323,6 +324,10 @@ protected boolean lessThan(IteratorAndCurrent<Bucket> a, IteratorAndCurrent<Buck
                     // the key changes, reduce what we already buffered and reset the buffer for current buckets
                     final Bucket reduced = reduceBucket(currentBuckets, reduceContext);
                     if (reduced.getDocCount() >= minDocCount || reduceContext.isFinalReduce() == false) {
+                        if (consumeBucketCount++ >= REPORT_EMPTY_EVERY) {
+                            reduceContext.consumeBucketsAndMaybeBreak(consumeBucketCount);
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalDateHistogram.java:328]: I think we should only do this is is final reduce?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalDateHistogram.java [hunk 2]
```diff
@@ -344,10 +349,14 @@ protected boolean lessThan(IteratorAndCurrent<Bucket> a, IteratorAndCurrent<Buck
                 final Bucket reduced = reduceBucket(currentBuckets, reduceContext);
                 if (reduced.getDocCount() >= minDocCount || reduceContext.isFinalReduce() == false) {
                     reducedBuckets.add(reduced);
+                    if (consumeBucketCount++ >= REPORT_EMPTY_EVERY) {
+                        reduceContext.consumeBucketsAndMaybeBreak(consumeBucketCount);
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalDateHistogram.java:353]: I think we should only do this is is final reduce?

**Review by iverase — COMMENTED** [review]
I am not sure if we should only call #consumeBucketsAndMaybeBreak for final reduce or we can rely that for partial reduce it should be a NOOP. Otherwise it is a good change. I want to hear your thought and I will approve it.

**Review by iverase — APPROVED** [review]
I look in other places and rely in the operation to be a noop for partial reduce.

**Comment by martijnvg** [comment]
> I look in other places and rely in the operation to be a noop for partial reduce.

Right in case of `ForPartial#consumeBucketCountAndMaybeBreak(...)` this is a noop.

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/part-3

**Pull request merged**